docs/spec/escrow.rst

```diff
@@ -0,0 +1,312 @@
+
+.. tabularcolumns:: |l|L|
+
+==============  ============  ========================================================	
+Parameter       Type          Description                                            
+==============  ============  ========================================================
+``bias``        Balance       The bias for the linear function.
+``slope``       Balance       The slope for the linear function.
+``ts``          BlockNumber   The current block height when this point was stored.
+==============  ============  ========================================================
+
+Functions
+~~~~~~~~~
+
+.. _escrow-function-create-lock:
+
+create_lock
+-----------
+
+Create a lock on the account's balance to expire in the future.
+
+Specification
+.............
+
+*Function Signature*
+
+``create_lock(who, amount, unlock_height)``
+
+*Parameters*
+
+* ``who``: The user's address.
+* ``amount``: The amount to be locked.
+* ``unlock_height``: The height to lock until.
+
+*Events*
+
+* :ref:`escrow-event-deposit`
+
+*Preconditions*
+
+* The function call MUST be signed by ``who``.
+* The ``amount`` MUST be non-zero.
+* The account ``who`` MUST NOT already have locked balance.
+* The ``unlock_height`` MUST be greater than ``now``.
+* The ``unlock_height`` MUST NOT be greater than ``now + MaxPeriod``.
+
+*Postconditions*
+
+* The account's ``LockedBalance`` MUST be set as follows:
+
+    * ``new_locked.amount``: MUST be the ``amount``.
+    * ``new_locked.end``: MUST be the ``unlock_height``.
+
+.. _escrow-function-increase-amount:
+
+increase_amount
+---------------
+
+Deposit additional tokens for a pre-existing lock to improve voting power.
+
+Specification
+.............
+
+*Function Signature*
+
+``increase_amount(who, amount)``
+
+*Parameters*
+
+* ``who``: The user's address.
+* ``amount``: The amount to be locked.
+
+*Events*
+
+* :ref:`escrow-event-deposit`
+
+*Preconditions*
+
+* The function call MUST be signed by ``who``.
+* The ``amount`` MUST be non-zero.
+* The account's ``old_locked.amount`` MUST be non-zero.
+* The account's ``old_locked.end`` MUST be greater than ``now``.
+
+*Postconditions*
+
+* The account's ``LockedBalance`` MUST be set as follows:
+
+    * ``new_locked.amount``: MUST be ``old_locked.amount + amount``.
+    * ``new_locked.end``: MUST be the ``old_locked.end``.
+
+.. _escrow-function-increase-unlock-height:
+
+increase_unlock_height
+----------------------
+
+Push back the expiry on a pre-existing lock to retain voting power.
+
+Specification
+.............
+
+*Function Signature*
+
+``increase_unlock_height(who, unlock_height)``
+
+*Parameters*
+
+* ``who``: The user's address.
+* ``unlock_height``: The new expiry deadline.
+
+*Events*
+
+* :ref:`escrow-event-deposit`
+
+*Preconditions*
+
+* The function call MUST be signed by ``who``.
+* The ``amount`` MUST be non-zero.
+* The account's ``old_locked.amount`` MUST be non-zero.
+* The account's ``old_locked.end`` MUST be greater than ``now``.
```

Comment:
I think we can lift that requirement. If the lock has passed, the escrow is not lifted until the user actually withdraws their tokens. That way the user can just call this function instead of having to lock from scratch again.

Reply:
+1

Reply:
This is as it is implemented in curve's contracts, I'm wary of removing any such checks as I'm not sure how it will affect the core logic.

Reply:
Ok, then let's leave it in